src/connstat.cpp

```diff
@@ -0,0 +1,169 @@
+    thrd_.detach();
+}
+
+void
+ConnectivityStatus::setTopicListener(connstat_event_cb ucb, topic topic)
+{
+    switch (topic) {
+    case topic::NEWADDR:
+    case topic::DELADDR:
+    case topic::ADDR:
+        nl_socket_add_memberships(nlsk.get(), RTNLGRP_IPV6_IFADDR, RTNLGRP_NONE);
+        nl_socket_add_memberships(nlsk.get(), RTNLGRP_IPV4_IFADDR, RTNLGRP_NONE);
+        break;
+    default:
+        break;
+    }
+    long unsigned int t = static_cast<long unsigned int>(topic);
+    topic_cbs[t] = ucb;
+}
+
+void
+ConnectivityStatus::removeTopicListener(topic topic)
+{
+    long unsigned int t = static_cast<long unsigned int>(topic);
+    switch (topic) {
+    case topic::NEWADDR:
+        nl_socket_drop_memberships(nlsk.get(), RTNLGRP_IPV6_IFADDR, RTNLGRP_NONE);
+        topic_cbs[t] = {};
+        break;
+    case topic::DELADDR:
+        nl_socket_drop_memberships(nlsk.get(), RTNLGRP_IPV4_IFADDR, RTNLGRP_NONE);
+        topic_cbs[t] = {};
+        break;
+    default:
+        break;
+    }
+}
+
+void
+ConnectivityStatus::executer(topic topic)
+{
+    long unsigned int t = static_cast<unsigned int>(topic);
+    if (topic_cbs[t])
+        (topic_cbs[t])(t);
+}
+
+bool
+ConnectivityStatus::nlsk_setup(nl_sock* nlsk)
+{
+    /*
+     * Notifications do not use sequence numbers, disable sequence number
+     * checking.
+     */
+    nl_socket_disable_seq_check(nlsk);
+
+    /*
+     * Define a callback function, which will be called for each notification
+     * received
+     */
+    nl_socket_modify_cb(nlsk, NL_CB_VALID, NL_CB_CUSTOM, nlcb, (void *)this);
+
+    /* Connect to routing netlink protocol */
+    nl_connect(nlsk, NETLINK_ROUTE);
+
+    /* TODO: return proper status */
+    return SUCCESS;
+}
+
+ConnectivityStatus::NlPtr
+ConnectivityStatus::nlsk_init(void)
+{
+    NlPtr ret(nl_socket_alloc(), &nl_socket_free);
+    if (not ret.get())
+        throw std::runtime_error("couldn't allocate netlink socket!\n");
+
+
+    return ret;
+}
+
+bool
+ConnectivityStatus::get_neigh_state(struct nl_msg* msg)
+{
+    struct nlmsghdr *h = nlmsg_hdr(msg);
+    struct ndmsg *r = (struct ndmsg*)nlmsg_data(h);
+
+    switch (r->ndm_state) {
+        case NUD_REACHABLE:
+            executer(topic::NEWNEIGH);
+            break;
+        default:
+            break;
+    }
+    return SUCCESS;
+}
+
+bool
+ConnectivityStatus::nl_event_cb(struct nl_msg* msg)
+{
+    struct nlmsghdr *h = nlmsg_hdr(msg);
+
+    switch (h->nlmsg_type) {
+        case RTM_NEWADDR:
+            executer(topic::ADDR);
+            executer(topic::NEWADDR);
+            break;
+        case RTM_DELADDR:
+            executer(topic::ADDR);
+            executer(topic::DELADDR);
+            break;
+        case RTM_NEWNEIGH:
+        case RTM_DELNEIGH:
+        case RTM_NEWNEIGHTBL:
+            return get_neigh_state(msg);
+            break;
+        default:
+            break;
+    }
+
+    return SUCCESS;
+}
+
+static int nlcb(struct nl_msg* msg, void* arg)
```

Comment:
no need to split declaration/definition (could even be inlined as a lambda in call to  nl_socket_modify_cb)